refactor: add multiprocessing, resilience, progress bar

Midnighter opened the pull request:
* Use `click` for the CLI and properly installs its command line tool.
* In addition to the above features also allows output of time in either days, hours, or minutes.
* Split up the code a bit more.

Ideas for the future:
* Allow analysis of individual trays/plates
* Move some Python loops to numpy

KristianJensen committed:
Update README.md

Midnighter committed:
feat: add progress, multiproc, and error catching

Midnighter committed:
refactor: add time units and conversion to OD

Midnighter committed:
docs: modify readme

Midnighter committed:
fix: improve flake8 and other style issues

KristianJensen commented on gp_align/align.py:
```diff
@@ -7,36 +26,40 @@
-    for i in np.linspace(-r, r, 2*r+1).astype(int):
-        for j in np.linspace(-r, r, 2*r+1).astype(int):
+    # TODO: Potentially use vectorized numpy functions here.
+    for i in np.arange(-r, r, 2 * r + 1):
```
This is not working. To work with arange it should be `arange(-r, r+1)`

Midnighter replied:
I saw this last moment and apparently introduced a bug right away :)

Midnighter replied:
On second thought, is there any reason not to use the normal `range` function?

KristianJensen replied:
I honestly don't remember why I went for linspace and type conversion. I don't see any reason why range would not work

KristianJensen commented on gp_align/align.py:
```diff
@@ -7,36 +26,40 @@
     best_value = 0
-    for i in np.linspace(-r, r, 2*r+1).astype(int):
-        for j in np.linspace(-r, r, 2*r+1).astype(int):
+    # TODO: Potentially use vectorized numpy functions here.
```
I guess a convolution could be used, but I don't know how much it would speed up as most of the time is spent doing `overlap = (image1_slice & image2_slice).sum()`

KristianJensen commented on gp_align/analysis.py:
```diff
@@ -0,0 +1,212 @@
+    analyze_image.config = config
+
+
+def calibrate_image(filename):
```
Not a very accurate function name. `detect_edges` or similar would be better.

KristianJensen commented on gp_align/analysis.py:
```diff
@@ -0,0 +1,212 @@
+def analyze_image(filename):
+    """Analyze all wells from all trays in one image."""
+    LOGGER.debug(filename)
+    config = analyze_image.config
```
This seems convoluted and slightly hacky. Is this a normal design pattern? Maybe it would be more concise to just pass the config dict as an argument to `analyze_image`

Midnighter replied:
It's not great. I agree. The point is that this function is basically the multiprocessing worker and the config always contains the same information. I'll take a look whether it has a huge impact to pass it in as an argument.

KristianJensen commented on gp_align/analysis.py:
```diff
@@ -0,0 +1,212 @@
+
+    for i, (plate_name, plate_image) in enumerate(
+            zip(config["plate_names"], plate_images)):
+        plate = data.setdefault(plate_name, dict())
```
Is setdefault necessary? As far as I can see this line will only be executed once for each plate_name. `plate = data[plate_name] = dict()` would be clearer imo

Midnighter replied:
Fair enough. I have no preference here.

KristianJensen commented on gp_align/conversion.py:
```diff
@@ -0,0 +1,57 @@
+        The transformed data frame with OD values.
+
+    """
+    wells = [col for col in df.columns if col.lower() != "time"]
```
It would be simpler to just let time be the index instead of a column. Then the conversion would just be `df = g2od(df, *params)`

Midnighter replied:
In that case simpler, true, but I'm not a fan at all of the output of the index without a name. I'll see if the index has a name whether that is also output to CSV.

KristianJensen replied:
You can give the index a name with `df.index.name`. This will be put in the top left "cell" of the csv

KristianJensen commented:
Great changes and additions!

Midnighter commented:
So what's up with AppVeyor here? Was it even used before to test the project?

KristianJensen commented:
No idea why AppVeyor is suddenly testing stuff (or what it's testing for that matter). There has not been any CI testing on this project previously

Midnighter committed:
fix: address Kristian's comments on GitHub

Midnighter committed:
fix: pass config as proper argument

Midnighter committed:
fix: use builtin range

Midnighter commented:
I think I addressed all your comments Kristian. I converted the README from .md to .rst so I don't know how to get rid of the conflict reported here. I may make another PR to add code for finding the parameters from a defined OD series but I consider this one done.

KristianJensen committed:
remove README.md
